src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,460 @@
+    RefreshTokenAuthorizerFactory,
+)
+
+
+def _default_filename(app_name: str) -> str:
+    r"""
+    construct the filename for the default JSONTokenStorage to use
+
+    on Windows, this is typically
+        ~\AppData\Local\globus\app\{app_name}/tokens.json
+
+    on Linux and macOS, we use
+        ~/.globus/app/{app_name}/tokens.json
+    """
+    if sys.platform == "win32":
+        # try to get the app data dir, preferring the local appdata
+        datadir = os.getenv("LOCALAPPDATA", os.getenv("APPDATA"))
+        if not datadir:
+            home = os.path.expanduser("~")
+            datadir = os.path.join(home, "AppData", "Local")
+        return os.path.join(datadir, "globus", "app", app_name, "tokens.json")
+    else:
+        return os.path.expanduser(f"~/.globus/app/{app_name}/tokens.json")
+
+
+@dataclass(frozen=True)
+class GlobusAppConfig:
+    """
+    Various configuration options for controlling the behavior of a ``GlobusApp``.
+
+    :param login_flow_manager: an optional ``LoginFlowManager`` instance or class.
+        An instance will be used directly when driving app login flows. A class will
+        be initialized with the app's ``login_client`` and this config's
+        ``request_refresh_tokens``.
+        If not given the default behavior will depend on the type of ``GlobusApp``.
+    :param token_storage: a ``TokenStorage`` instance or class that will
+        be used for storing token data. If not passed a ``JSONTokenStorage``
+        will be used.
+    :param request_refresh_tokens: If True, the ``GlobusApp`` will request refresh
+        tokens for long lived access.
+    """
+
+    login_flow_manager: LoginFlowManager | type[LoginFlowManager] | None = None
+    token_storage: TokenStorage | None = None
+    request_refresh_tokens: bool = False
+
+
+class GlobusApp(metaclass=abc.ABCMeta):
+    """
+    ``GlobusApp` is an abstract base class providing an interface for simplifying
+    authentication with Globus services.
+
+    A ``GlobusApp`` manages scope requirements across multiple resource servers,
+    runs login flows through its ``LoginFlowManager``, handles token storage and
+    validation through its ``ValidatingTokenStorage``, and provides up-to-date
+    authorizers from its ``AuthorizerFactory`` through ``get_authorizer``.
+
+    In the near future, a ``GlobusApp`` will be accepted as an initialization parameter
+    to any Globus service client, automatically handling the client's default
+    scope requirements and providing the client an authorizer.
+    """
+
+    _login_client: AuthLoginClient
+    _authorizer_factory: AuthorizerFactory[GlobusAuthorizer]
+
+    def __init__(
+        self,
+        app_name: str,
+        *,
+        login_client: AuthLoginClient | None = None,
+        client_id: UUIDLike | None = None,
+        client_secret: str | None = None,
+        scope_requirements: dict[str, list[Scope]] | None = None,
+        config: GlobusAppConfig | None = None,
```

Comment:
With config being immutable now, we should be able to set the default in the constructor.

```suggestion
        config: GlobusAppConfig  = GlobusAppConfig()
```